pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,262 @@
+			return true
+		},
+	)
+
+	cacherefreshlog.Info("Aligning ResourceOffers - PeeringInfo")
+	if err := checkAlignmentResourceOfferPeeringInfo(ctx, spv); err != nil {
+		return true, err
+	}
+	cacherefreshlog.Info("Cache refresh completed")
+	return false, nil
+}
+
+func alignTerminatingOrNotExistingShadowPods(shadowPodList *vkv1alpha1.ShadowPodList, pi *peeringInfo) {
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
+	spMap := make(map[string]string)
+	cacherefreshlog.Info(fmt.Sprintf("Found %d ShadowPods for cluster %q", len(shadowPodList.Items), pi.clusterName))
+	// Check on all cluster ShadowPods and saving them in a temporary map
+	checkShadowPodsAlignment(shadowPodList, pi, spMap)
+	cacherefreshlog.Info("Searching for terminated ShadowPodDescription to be removed from cache")
+	// Alignment of all ShadowPodDescriptions in cache
+	alignTerminatingShadowPodDescriptions(pi, spMap)
+}
+
+func alignTerminatingShadowPodDescriptions(pi *peeringInfo, spMap map[string]string) {
+	for _, shadowPodDescription := range pi.spDescriptionList {
+		// Check if the ShadowPod is in terminating phase in cache and has been already terminated/deleted from the cluster
+		// if true ShadowPodDescription can be also deleted from the cache
+		if _, stillPresent := spMap[shadowPodDescription.namespacedName.String()]; !shadowPodDescription.running && !stillPresent {
+			pi.removeShadowPod(shadowPodDescription)
+			cacherefreshlog.Info(fmt.Sprintf("ShadowPodDescription %s removed from cache", shadowPodDescription.namespacedName.String()))
+		}
+	}
+}
+
+func alignExistingShadowPods(shadowPodList *vkv1alpha1.ShadowPodList, pi *peeringInfo) {
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
+	cacherefreshlog.Info(fmt.Sprintf("Found %d ShadowPods running on cluster %s", len(shadowPodList.Items), pi.clusterName))
+	for i := range shadowPodList.Items {
+		shadowPod := &shadowPodList.Items[i]
+		nsname := types.NamespacedName{Name: shadowPod.Name, Namespace: shadowPod.Namespace}
+		_, found := pi.spDescriptionList[nsname.String()]
+		if !found {
+			// TODO: check if it is the correct logic
+			// Errors are intentionally ignored here. Pod is already running on cluster.
+			spQuota, _ := getQuotaFromShadowPod(shadowPod, false)
+			pi.addShadowPod(createShadowPodDescription(shadowPod.GetName(), shadowPod.GetNamespace(), shadowPod.GetUID(), spQuota))
+			cacherefreshlog.Info(fmt.Sprintf("ShadowPod %s added in cache", nsname.String()))
+		}
+	}
+}
+
+func checkShadowPodsAlignment(shadowPodList *vkv1alpha1.ShadowPodList, pi *peeringInfo, spMap map[string]string) {
+	for i := range shadowPodList.Items {
+		shadowPod := &shadowPodList.Items[i]
+		nsname := types.NamespacedName{Name: shadowPod.Name, Namespace: shadowPod.Namespace}
+		_, found := pi.spDescriptionList[nsname.String()]
+		if !found {
+			cacherefreshlog.Info(fmt.Sprintf("Warning: ShadowPod %s not found in cache", nsname.String()))
+			// TODO: this should never happen, but maybe we need to manage also this case better
```

Comment:
This might happen in case the webhook is not configured in fail mode. It would be probably appropriate to add that sp to the cache and align the resources.

Reply:
and also validate it before?

Reply:
No, given the shadowpod already exists.